### PR TITLE
Fix ApplicableSpan calculation

### LINKED ISSUE #23: Exception `Start and End must be well ordered`
https://github.com/Microsoft/vscode-python/issues/2221

```python
a = [1, "str", 3]
a[0]. #<- exception here
```

image

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
Fixes #23 

- With whitespace surrounding dot span of the expression can be to the right of the dot. Adjust span accordingly.
- Add test for the case (reproes only when completion context is supplied)

### CODE REVIEW

File: src/LanguageServer/Impl/Implementation/Server.Completion.cs [hunk 1]
```diff
@@ -38,7 +40,7 @@ public sealed partial class Server {
             var analysis = entry != null ? await entry.GetAnalysisAsync(50, cancellationToken) : null;
             if (analysis == null) {
                 TraceMessage($"No analysis found for {uri}");
-                return new CompletionList();
+                return EmptyCompletion;
```
Comment (AlexanderSher) [inline src/LanguageServer/Impl/Implementation/Server.Completion.cs]: It is not safe to return shared instance cause `CompletionList` isn't immutable,

Comment (MikhailArkhipov) [reply src/LanguageServer/Impl/Implementation/Server.Completion.cs]: OK

Comment (MikhailArkhipov) [reply src/LanguageServer/Impl/Implementation/Server.Completion.cs]: We do return hover though

Comment (AlexanderSher) [reply src/LanguageServer/Impl/Implementation/Server.Completion.cs]: Considering amount of objects created by Json.Net, another instance won't make a big difference.

File: src/Analysis/Engine/Test/CompletionTests.cs [hunk 2]
```diff
@@ -1003,6 +1003,15 @@ class Simple(unittest.TestCase):
             }
         }
 
+        [TestMethod, Priority(0)]
+        public async Task WithWhitespaceAroundDot() {
+            using (var s = await CreateServerAsync()) {
+                var u = await s.OpenDefaultDocumentAndGetUriAsync("import sys\nsys  .  version\n");
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Test/CompletionTests.cs]: Do we care about `\r\n`?

File: src/LanguageServer/Impl/Implementation/Server.Completion.cs [hunk 3]
```diff
@@ -70,39 +69,50 @@ public sealed partial class Server {
...
+                return res;
+            }
+
+            await InvokeExtensionsAsync((ext, token)
+                => (ext as ICompletionExtension)?.HandleCompletionAsync(uri, analysis, tree, @params.position, res, cancellationToken), cancellationToken);
+
+            return res;
+        }
+
+        private SourceSpan? GetApplicableSpan(CompletionAnalysis ca, CompletionParams @params, PythonAst tree) {
```
Comment (AlexanderSher) [inline src/LanguageServer/Impl/Implementation/Server.Completion.cs]: Does any of it affects `textEdit` field in the individual `CompletionItem`?

File: src/Analysis/Engine/Test/CompletionTests.cs (hunk 2 above)
Comment (MikhailArkhipov) [reply src/Analysis/Engine/Test/CompletionTests.cs]: Not here, I believe there are parser tests for the line endings

File: src/LanguageServer/Impl/Implementation/Server.Completion.cs (hunk 3 above)
Comment (MikhailArkhipov) [reply src/LanguageServer/Impl/Implementation/Server.Completion.cs]: Technically they should all have the same applicable span. The span is actually used only by VS to figure out what to delete and replace with the completion item insertion text.

**Pull request merged**